[Cocoa] Temporarily wake up remote process if outgoing IPC message queue becomes too large

cdumez opened the pull request:
#### 51833439139b0bb53857d4342ce05b61dfde1683
<pre>
[Cocoa] Temporarily wake up remote process if outgoing IPC message queue becomes too large
<a href="https://bugs.webkit.org/show_bug.cgi?id=256181">https://bugs.webkit.org/show_bug.cgi?id=256181</a>

Reviewed by Ben Nham.

Temporarily wake up remote process if outgoing IPC message queue becomes too
large, to avoid potentially bad memory growth when the remote process is
suspended.

This currently only works for AuxiliaryProcessProxy, so IPC from the UIProcess
to child processes. However, we may want to consider extending this to IPC
from the network process to WebProcesses in a follow-up.

In the current proposal, we wake up the destination process for 3 seconds when
its outgoing message queue has reached 1024 messages. We do so at most once
every 30 seconds.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::setOutgoingMessageQueueIsGrowingLargeCallback):
(IPC::Connection::invalidate):
(IPC::Connection::sendMessage):
* Source/WebKit/Platform/IPC/Connection.h:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::AuxiliaryProcessProxy):
(WebKit::AuxiliaryProcessProxy::didFinishLaunching):
(WebKit::AuxiliaryProcessProxy::outgoingMessageQueueIsGrowingLarge):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:

Canonical link: <a href="https://commits.webkit.org/263570@main">https://commits.webkit.org/263570@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/966aea09d368713f0511210d348725363b8f1913

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5086 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5217 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5394 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6607 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5173 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5424 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5193 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/5420 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5173 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5267 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4561 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6623 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2754 "Passed tests") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/9700 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4617 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4633 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6241 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5051 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4146 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4534 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1218 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8614 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4899 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -55,9 +55,12 @@ namespace IPC {
+constexpr size_t maxPendingIncomingMessagesKillingThreshold { 50000 };
 #endif
 
+constexpr size_t largeOutgoingMessageQueueCountThreshold { 500 };
```
I'd suggest picking 1024 to match `MACH_PORT_QLIMIT_LARGE`. So basically in the worst case this will trigger after ~2048 messages if the receiver process is suspended (1024 will be stuck in the Mach IPC queue and 1024 will be stuck in `m_outgoingMessages`.

I also realize this only applies to UIProcess at the moment, but when running some LocalStorage benchmarks, I was easily able to see message queues of size ~3k-4k in NetworkProcess even without the receiver process being suspended. So we may need to consider upping this limit so that it doesn't become too noisy. But I think we can start with this limit.

cdumez replied:
Will up to 1024 for now. We can up if we becomes too noisy later. But it won't be more than once every 30 seconds so I am not too worried about noise at the moment.

bnham commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -550,6 +559,17 @@ bool Connection::sendMessage(UniqueRef<Encoder>&& encoder, OptionSet<SendOption>
+                , 0
+#endif
+            );
+            m_outgoingMessageQueueIsGrowingLargeCallback();
```
Maybe this should be called outside of the block that holds the lock?

cdumez replied:
Ok.

bnham commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
```diff
@@ -311,6 +318,15 @@ void AuxiliaryProcessProxy::didFinishLaunching(ProcessLauncher*, IPC::Connection
+#if USE(RUNNINGBOARD)
+    // If we keep trying to send IPC to a suspended process, the outgoing message queue may grow large and result
+    // in increased memory usage. To avoid this, we wake up the process for a bit so we can drain the messages.
+    m_timedActivityForIPC = throttler().backgroundActivity("IPC sending due to large outgoing queue"_s);
```
This gets called back on the IPC send queue, does it need to dispatch to the main thread?

bnham replied:
Oh wait, never mind, seems like this gets called on the thread that tried to send the IPC. Which I guess in most cases is the main thread. Are there cases where we try to send IPCs off the main thread in UIProcess?

cdumez replied:
This doesn't get called on the IPC send queue, it gets called on the same thread as `send()`. That said, it is true that `send()` can be called off the main thread. I'll refine.

bnham commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -547,9 +556,26 @@ bool Connection::sendMessage(UniqueRef<Encoder>&& encoder, OptionSet<SendOption>
+    }
+
+    if (shouldNotifyOfQueueGrowingLarge) {
+        RELEASE_LOG_ERROR(IPC, "Connection::sendMessage(): Too many messages (%lu) in the queue to remote PID: %d, notifying client", outgoingMessagesCount
```
nit: `%zu` for `size_t`

bnham reviewed (approved):
LGTM

cdumez committed:
[Cocoa] Temporarily wake up remote process if outgoing IPC message queue becomes too large
https://bugs.webkit.org/show_bug.cgi?id=256181

Reviewed by Ben Nham.

Temporarily wake up remote process if outgoing IPC message queue becomes too
large, to avoid potentially bad memory growth when the remote process is
suspended.

This currently only works for AuxiliaryProcessProxy, so IPC from the UIProcess
to child processes. However, we may want to consider extending this to IPC
from the network process to WebProcesses in a follow-up.

In the current proposal, we wake up the destination process for 3 seconds when
its outgoing message queue has reached 1024 messages. We do so at most once
every 30 seconds.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::setOutgoingMessageQueueIsGrowingLargeCallback):
(IPC::Connection::invalidate):
(IPC::Connection::sendMessage):
* Source/WebKit/Platform/IPC/Connection.h:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::AuxiliaryProcessProxy):
(WebKit::AuxiliaryProcessProxy::didFinishLaunching):
(WebKit::AuxiliaryProcessProxy::outgoingMessageQueueIsGrowingLarge):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:

Canonical link: https://commits.webkit.org/263570@main
